metals/src/main/scala/scala/meta/internal/metals/callHierarchy/FindCallsResult.scala

```diff
@@ -0,0 +1,158 @@
+import scala.meta.io.AbsolutePath
+import scala.meta.internal.semanticdb.TextDocument
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.concurrent.Future
+import scala.meta.pc.CancelToken
+import scala.jdk.CollectionConverters._
+import scala.concurrent.ExecutionContext
+import java.{util => ju}
+
+private[callHierarchy] sealed trait FindCallsResult[T] {
+  def occurence: SymbolOccurrence
+  def definitionNameRange: lsp4j.Range
+  def fromRanges: List[lsp4j.Range]
+
+  protected def toLSP(
+      lspBuilder: (lsp4j.CallHierarchyItem, ju.List[lsp4j.Range]) => T
+  )(source: AbsolutePath, doc: TextDocument)(
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  )(implicit ec: ExecutionContext): Future[T] =
+    occurence match {
+      case SymbolOccurrence(_, symbol, _) =>
+        builder
+          .build(
+            source,
+            doc,
+            occurence,
+            definitionNameRange,
+            visited :+ symbol,
+            token,
+          )
+          .collect { case Some(item) =>
+            lspBuilder(
+              item,
+              fromRanges.asJava,
+            )
+          }
+    }
+}
+
+/**
+ * Result of incoming calls finder.
+ *
+ * @param canBuildHierarchyItem The item that makes the call wrapped in CanBuildHierarchyItem.
+ * @param definitionNameRange The range that should be selected.
+ * @param fromRanges Ranges at which the calls appear.
+ */
+private[callHierarchy] case class FindIncomingCallsResult(
+    occurence: SymbolOccurrence,
+    definitionNameRange: lsp4j.Range,
+    fromRanges: List[lsp4j.Range],
+)(implicit ec: ExecutionContext)
+    extends FindCallsResult[lsp4j.CallHierarchyIncomingCall] {
+
+  private val anonToLSP: (AbsolutePath, TextDocument) => (
+      CallHierarchyItemBuilder,
+      Array[String],
+      CancelToken,
+  ) => Future[lsp4j.CallHierarchyIncomingCall] =
+    toLSP((from, fromRanges) =>
+      new lsp4j.CallHierarchyIncomingCall(from, fromRanges)
+    ) _
+
+  def toLSP(
+      source: AbsolutePath,
+      doc: TextDocument,
+      builder: CallHierarchyItemBuilder,
+      visited: Array[String],
+      token: CancelToken,
+  ) =
+    anonToLSP(source, doc)(builder, visited, token)
+}
+
+private[callHierarchy] object FindIncomingCallsResult {
+
+  /** Aggregate results of incoming calls finder by grouping the fromRanges. */
+  def group[T](
+      results: List[FindIncomingCallsResult]
+  )(implicit ec: ExecutionContext): List[FindIncomingCallsResult] =
+    results.groupBy(_.occurence).values.toList.map {
+      case result @ FindIncomingCallsResult(
+            occurence,
+            definitionNameRange,
+            _,
+          ) :: _ =>
+        FindIncomingCallsResult(
+          occurence,
+          definitionNameRange,
+          result.flatMap(_.fromRanges),
+        )
+      case _ =>
+        throw new IllegalArgumentException("Can't apply group on a empty list.")
```

Comment:
I don't think we should throw here. If we hit this maybe log a warning if you want and then return an empty list.